validator/src/bootstrap.rs

```diff
@@ -369,6 +369,16 @@ pub fn fail_rpc_node(
     blacklisted_rpc_nodes.insert(*rpc_id);
 }
 
+fn shutdown_gossip_service(
+    gossip: &mut Option<(Arc<ClusterInfo>, Arc<AtomicBool>, GossipService)>,
+) {
+    if let Some((cluster_info, gossip_exit_flag, gossip_service)) = gossip.take() {
+        cluster_info.save_contact_info();
+        gossip_exit_flag.store(true, Ordering::Relaxed);
+        gossip_service.join().unwrap();
+    }
+}
```

Comment:
It currently needs to be mutable because of the take. It could be reworked like so:

fn caller() {
    if let Some(gossip) = gossip {
        shutdown_gossip_service(gossip);
    }
}

fn shutdown_gossip_service(
    gossip: (Arc<ClusterInfo>, Arc<AtomicBool>, GossipService),
) {
    let (cluster_info, gossip_exit_flag, gossip_service) = gossip;
    cluster_info.save_contact_info();
    gossip_exit_flag.store(true, Ordering::Relaxed);
    gossip_service.join().unwrap();
}

Reply:
#30986 to follow up on this